paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentOptionsViewModel.kt

```diff
@@ -336,15 +334,52 @@ internal class PaymentOptionsViewModel @Inject constructor(
         ): T {
             val application = applicationSupplier()
             val starterArgs = starterArgsSupplier()
-            injectWithFallback(
-                starterArgsSupplier().injectorKey,
-                FallbackInitializeParam(application, starterArgs.productUsage)
-            )
-            return subComponentBuilderProvider.get()
+            val logger = Logger.getInstance(BuildConfig.DEBUG)
+
+            WeakMapInjectorRegistry.retrieve(starterArgs.injectorKey)?.let {
+                it as? NonFallbackInjector
+            }?.let {
+                logger.info(
+                    "Injector available, " +
+                        "injecting dependencies into ${this::class.java.canonicalName}"
+                )
+                injector = it
+                it.inject(this)
+            } ?: run {
+                logger.info(
+                    "Injector unavailable, " +
+                        "initializing dependencies of ${this::class.java.canonicalName}"
+                )
+                fallbackInitialize(FallbackInitializeParam(application, starterArgs.productUsage))
+            }
+
+            val subcomponent = subComponentBuilderProvider.get()
                 .application(application)
                 .args(starterArgs)
                 .savedStateHandle(savedStateHandle)
-                .build().viewModel as T
+                .build()
+            val viewModel = subcomponent.viewModel
+            viewModel.injector = injector
+            return viewModel as T
+        }
+
+        override fun fallbackInitialize(arg: FallbackInitializeParam) {
+            val component = DaggerPaymentOptionsViewModelFactoryComponent.builder()
+                .context(arg.application)
+                .productUsage(arg.productUsage)
+                .build()
+
+            injector = object : NonFallbackInjector {
+                override fun inject(injectable: Injectable<*>) {
+                    when (injectable) {
+                        is FormViewModel.Factory -> component.inject(injectable)
+                        else -> {
+                            throw IllegalArgumentException("invalid Injectable $injectable requested in $this")
+                        }
+                    }
+                }
+            }
+            component.inject(this)
```

Comment:
This logic is now replicated in `PaymentOptionsViewModel`, `PaymentSheetViewModel` and `LinkActivityViewModel`. I'll refactor to make it generic and part of the `stripe-core` module in the next PR.

Reply:
#5739